src/pages/trade/store/positions.ts

```diff
@@ -0,0 +1,99 @@
+import { makeAutoObservable } from 'mobx';
+import { TransactionPlannerRequest } from '@penumbra-zone/protobuf/penumbra/view/v1/view_pb';
+import { PositionId } from '@penumbra-zone/protobuf/penumbra/core/component/dex/v1/dex_pb';
+import { planBuildBroadcast } from '@/pages/trade/ui/order-form/helpers.tsx';
+import { connectionStore } from '@/shared/model/connection';
+import { AddressIndex } from '@penumbra-zone/protobuf/penumbra/core/keys/v1/keys_pb';
+import { queryClient } from '@/shared/const/queryClient.ts';
+import { getPositionData, PositionData } from '@/pages/trade/api/positions.ts';
+import { penumbra } from '@/shared/const/penumbra.ts';
+import { DexService } from '@penumbra-zone/protobuf';
+import { openToast } from '@penumbra-zone/ui/Toast';
+
+class PositionsStore {
+  public loading = false;
+
+  constructor() {
+    makeAutoObservable(this);
+  }
+
+  setLoading(state: boolean) {
+    this.loading = state;
+  }
+
+  closePositions = async (positions: PositionId[]): Promise<void> => {
+    try {
+      this.setLoading(true);
+
+      const planReq = new TransactionPlannerRequest({
+        positionCloses: positions.map(positionId => ({ positionId })),
+        source: new AddressIndex({ account: connectionStore.subaccount }),
+      });
+
+      await planBuildBroadcast('positionClose', planReq);
+      await this.updatePositionsInCache(positions);
+    } catch (e) {
+      openToast({
+        type: 'error',
+        message: 'Error with withdraw action',
+        description: String(e),
+      });
+    } finally {
+      this.setLoading(false);
+    }
+  };
+
+  withdrawPositions = async (positions: PositionId[]): Promise<void> => {
+    try {
+      this.setLoading(true);
+
+      // Fetching latest position data as the planner request requires current reserves + pair
+      const promises = positions.map(positionId =>
+        penumbra.service(DexService).liquidityPositionById({ positionId }),
+      );
+      const latestPositionData = await Promise.all(promises);
+
+      const planReq = new TransactionPlannerRequest({
+        positionWithdraws: positions.map((positionId, i) => ({
+          positionId,
+          tradingPair: latestPositionData[i]?.data?.phi?.pair,
+          reserves: latestPositionData[i]?.data?.reserves,
+        })),
+        source: new AddressIndex({ account: connectionStore.subaccount }),
+      });
+
+      await planBuildBroadcast('positionWithdraw', planReq);
```

Comment:
suggestion: `planBuildBroadcast` may fail, can we add an [errorToast](https://github.com/penumbra-zone/web/blob/main/apps/minifront/src/state/ibc-out.ts#L111)? The subsequent `updatePositionsInCache` could also fail.